ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/ConnectionManager.hs

```diff
@@ -0,0 +1,438 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Test.Ouroboros.Network.Testnet.Simulation.ConnectionManager where
+
+import           Control.Monad.IOSim (Trace)
+
+import           Data.Maybe (fromJust, isJust)
+import           Data.Bitraversable (bimapAccumL)
+import           Data.List (find, dropWhileEnd)
+import qualified Data.List.Trace as Trace
+import qualified Data.Map as Map
+import           Data.Monoid (Sum(..))
+
+import           Text.Printf (printf)
+
+import           Ouroboros.Network.ConnectionManager.Types
+                     (AbstractTransition, Transition' (..), AbstractState (..),
+                     Provenance (..), DataFlow (..), TimeoutExpired (..),
+                     ConnectionManagerTrace (TrPruneConnections),
+                     AbstractTransitionTrace, TransitionTrace' (..))
+import           Ouroboros.Network.ConnectionHandler (ConnectionHandlerTrace)
+
+import           Test.QuickCheck
+                     (Property, (.&&.), Testable (property), label, tabulate,
+                     cover, counterexample)
+
+-- | Split 'AbstractTransitionTrace' into seprate connections.  This relies on
+-- the property that every connection is terminated with 'UnknownConnectionSt'.
+-- This property is verified by 'verifyAbstractTransitionOrder'.
+--
+splitConns :: Ord addr
+           => (a -> AbstractTransitionTrace addr)
+           -> Trace r a
+           -> Trace r [a]
```

Comment:
Any idea how could we avoid code duplication.  This will be especially annoying for `verifyAbstractTransition`.   Can we use a public library?

I wrote a comment about this somewhere else, earlier this week ...

Reply:
Another possible solution is to have `ouroboros-network-test` package which depends on both `ouroboros-network` and `ouroboros-network-framework`.

Reply:
Done!

Reply:
I thought we were supposed to move it to a testing library in `ouroboros-network-framework`, the same with `verifyAbstractTransitions`.

Reply:
Accidentally deleted this work, working on it